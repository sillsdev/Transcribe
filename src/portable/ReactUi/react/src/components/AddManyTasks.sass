@import "../_variables.sass"

.AddManyTasks
    width: 100%
    height: 100%
    position: absolute
    top: 0
    left: 0
    background-color:transparent
    display: flex
    flex-direction: row
    justify-content: flex-end
    &.rtl
        direction: rtl
        .panel
            padding-right: 0
            &>*
                padding-left: 0
            &>.titleRow
                padding-right: 0px
    .panel
        width: 510px
        height: 100%
        background-color: $base02
        display: flex
        flex-direction: column
        flex: 0 0 auto
        overflow-y: visible
        &>.titleRow
            padding-left: 0
            padding-bottom: 32px
        padding-left: 0
    .titleRow
        display: flex
        flex-direction: row
        padding-top: 32px
        z-index: 2
        .BackLink
            width: 64px
            text-align: center
    .NextAction
        display: flex
        font-family: $sans
        margin: auto
        cursor: pointer
        width: 100px
        padding: 0px 5px
        align-self: center

    .helpText
        color: $base1
        font-family: $sans
        font-size: 16px
        line-height: 28px
        letter-spacing: 0.44px
        padding: 24px 48px
        text-align: center

    .linkText
        color: $blueLink
        &:hover
            cursor: pointer

    .LinkAction
        width: 220px
        margin: 0px

    .browseDiv
        width: 446px
        height: 124px
        display: flex
        border: 1px dotted white
        justify-content: center
        vertical-align: middle
        margin-left: 32px
        margin-right: 32px
        align-items: center
