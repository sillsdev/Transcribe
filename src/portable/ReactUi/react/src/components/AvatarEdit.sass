@import "../_variables.sass"

.AvatarEdit
	width: 100%
	height: 100%
	position: absolute
	top: 0
	left: 0
	background-color: transparent
	display: -ms-flexbox
	display: flex
	-ms-flex-direction: row
	flex-direction: row
	-ms-flex-pack: end
	justify-content: flex-end
	&.rtl
        direction: rtl
        .panel
            padding-right: 0
            &>*
                padding-left: 0
                padding-right: 64px
            &>.titleRow
                padding-right: 0px
	.panel
        width: 352px
        height: 100%
        background-color: $base02
        display: flex
        flex-direction: column
        flex: 0 0 auto
        overflow-y: auto
        &>*
            padding-left: 64px
            flex: 0 0 auto
        &>.titleRow
            padding-left: 0
            padding-bottom: 32px
        padding-left: 0
	.titleRow
        display: flex
        flex-direction: row
        padding-top: 32px
        z-index: 2
        .BackLink
            width: 64px
            text-align: center
        .title
            align-self: flex-start
            flex: 1 0 auto
        .copyAction.hide
            display: none
        .copyAction
            display: flex
            flex-direction: column
            align-self: center
            padding: 0 64px
	.data
		z-index: 2
	.data>*>*
		width: 448px
	.dataColumn
		max-width: 392px
	.data>*>.SelectField .dataColumn
		max-width: 448px
	.options
		display: flex
		flex-direction: column
		flex: 1 0 auto
		padding: 70px
		.uploadContent
			position: relative
			overflow: hidden
			display: inline-block
		.btn
			border: 2px solid gray
			color: #002B36
			background-color: #F5CC4C
			padding: 8px 20px
			border-radius: 4px
			font-family: "Noto Sans", "Roboto", sans-serif
			font-size: 14px
			line-height: 16px
			letter-spacing: 0.75px
		.uploadContent input[type=file]
			font-size: 100px
			position: absolute
			left: 0
			top: 0
			opacity: 0
		.zoomContent
			padding-top: 20px
	.action
        background-color: $base02
        display: flex
        flex: 0 0 auto
        z-index: 1
        justify-content: space-between
        flex-direction: row
        border-top: 1px solid $base01
	.IconButtonField
		margin: 0
	#discard label
		width: 144px
input
	color: $base1
	font-family: $sans
	font-style: normal
	font-size: 12px
	line-height: 16px
	letter-spacing: 15%

