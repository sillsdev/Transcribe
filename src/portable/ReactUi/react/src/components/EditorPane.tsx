import * as React from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import * as actions2 from '../actions/audioActions';
// import { log } from '../actions/logAction';
import * as actions from '../actions/taskActions';
import { IState } from '../model/state';
import language from '../selectors/language';
import './EditorPane.sass';


interface IProps extends IStateProps, IDispatchProps {
}

const initialState = {
    seconds: 0,
    text: "",
}
class EditorPane extends React.Component<IProps, typeof initialState> {
    public state: typeof initialState;
    private interval: any;
    private textArea:any;

    public constructor(props: IProps) {
        super(props);
        this.state = {...initialState}
        this.change = this.change.bind(this);
        this.keyDown = this.keyDown.bind(this);
        this.textArea = React.createRef();
        this.blur = this.blur.bind(this);
    }

    public componentDidUpdate(){
        this.textArea.focus();
    }

    public componentDidMount() {
        this.interval = setInterval(() => this.tick(), 1000);
      }

    public componentWillUnmount() {
        clearInterval(this.interval);
    }

    public render() {
        const { direction, initialTranscription, saved, submit, lang, users,
            selectedProject, selectedTask, selectedUser, totalSeconds,
            transcription, writeTranscription } = this.props;
        const user = users.filter(u => u.username.id === selectedUser)[0];
        const project = user && user.project && user.project.filter(p => p.id === selectedProject)[0];
        const font = project != null? project.fontfamily: "SIL Charis"; // Tests null or undefined
        const size = project != null? project.fontsize: "12pt"; // Tests null or undefined
        const readOnlyStatus = (selectedTask !== undefined && selectedTask.length > 0)? false : true;
        // log("EditorPane")
        if (transcription != null && this.state.text !== transcription  && initialTranscription) {
            this.setState({text: transcription})
        }
        if ((selectedTask !== undefined && selectedTask.length > 0) && submit && !saved) {
            writeTranscription(selectedTask, totalSeconds, lang, direction?direction:"ltr", this.state)
        }
        return (
            <div className="EditorPane">
                <textarea
                    id="Editor"
                    dir={direction?direction:"ltr"}
                    ref={(textArea) => this.textArea = textArea}
                    value={this.state.text}
                    style={{fontFamily: font, fontSize: size}} 
                    onChange={this.change} 
                    onKeyDown={this.keyDown}
                    onFocus={this.movePositionAtEnd}
                    onBlur={this.blur}
                    readOnly={readOnlyStatus}
                    onMouseLeave={this.blur}
                />
            </div>
        )
    }

    public componentWillUpdate() {
        const { direction, lang, saved, selectedTask, totalSeconds, writeTranscription } = this.props;
        if (this.state.seconds === 60) {
            this.setState({
                seconds: 0
            });
            if ((selectedTask !== undefined && selectedTask.length > 0) && !saved) {
                writeTranscription(selectedTask, totalSeconds, lang, direction?direction:"ltr", this.state)
            }
        }
    }
    private tick() {
        this.setState(prevState => ({
            ...this.state,
            seconds: prevState.seconds + 1
        }));
    }

    private change(event: any) {
        this.setState({
            ...this.state,
            text: event.target.value
        });
    }

    private keyDown(event: any) {
        const { direction, setSaved, setInitialTranscription, lang, requestPosition, saved, selectedTask, totalSeconds, writeTranscription } = this.props;
        if ((selectedTask !== undefined && selectedTask.length > 0) && event.keyCode === 32) {
            requestPosition();
            writeTranscription(selectedTask, totalSeconds, lang, direction?direction:"ltr", this.state)
        } else if (saved) {
            setSaved(false);
        }
        setInitialTranscription(false);
    }

    private blur(event: any) {
        const { direction, lang, requestPosition, saved, selectedTask, totalSeconds, writeTranscription } = this.props;
        if((selectedTask !== undefined && selectedTask.length > 0) && !saved)
        {
            requestPosition();
            writeTranscription(selectedTask, totalSeconds, lang, direction?direction:"ltr", this.state)
        }
    }

    private movePositionAtEnd(e: any) {
        const tempValue = e.target.value
        e.target.value = ''
        e.target.value = tempValue
    }
};

interface IStateProps {
    direction?: string;
    initialTranscription: boolean;
    submit: boolean;
    lang: string;
    saved: boolean;
    selectedUser: string;
    selectedProject: string;
    selectedTask: string;
    totalSeconds: number;
    transcription: string;
    users: IUser[];
};

const mapStateToProps = (state: IState): IStateProps => ({
    direction: language(state).direction,
    initialTranscription: state.audio.initialTranscription,
    lang: language(state).lang,
    saved: state.audio.saved,
    selectedProject: state.tasks.selectedProject,
    selectedTask: state.tasks.selectedTask,
    selectedUser: state.users.selectedUser,
    submit: state.audio.submit,
    totalSeconds: state.audio.totalSeconds,
    transcription: state.audio.transcription,
    users: state.users.users,
});

interface IDispatchProps {
    setSaved: typeof actions2.setSaved;
    requestPosition: typeof actions2.requestPosition;
    setInitialTranscription: typeof actions2.setInitialTranscription;
    writeTranscription: typeof actions.writeTranscription;
};

const mapDispatchToProps = (dispatch: any): IDispatchProps => ({
    ...bindActionCreators({
        requestPosition: actions2.requestPosition,
        setInitialTranscription: actions2.setInitialTranscription,
        setSaved: actions2.setSaved,
        writeTranscription: actions.writeTranscription,
    }, dispatch),
});

export default connect(mapStateToProps, mapDispatchToProps)(EditorPane);
