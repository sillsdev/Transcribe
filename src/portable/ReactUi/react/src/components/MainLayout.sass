@import "../_variables.sass"

.MainLayout.Modal
	background-color: $base03
	-webkit-filter: blur(5px)
	-moz-filter: blur(5px)
	-o-filter: blur(5px)
	-ms-filter: blur(5px)
	filter: blur(5px)

.MainLayout 
	display: flex
	flex-direction: row
	justify-content: stretch
	align-items: stretch
	height: 100%
	width: 100%
	position: absolute
	overflow-y: hidden
	.NavCol 
		box-shadow: 0px 4px 8px rgba(0,0,0, 0.25)
		z-index: 3
		flex: 0 0 auto
	
	.TaskCol 
		box-shadow: 2px 4px 8px rgba(0,0,0, 0.35)
		z-index: 2
		flex: 0 0 auto
	
	.MainCol 
		flex: 1 0 auto
	
	/* Undo bootstrap box-sizing */
	* 
		box-sizing: unset
		-moz-box-sizing: unset
		-webkit-box-sizing: unset
	&.rtl
		direction: rtl

	.TodoBorderOn
		border: 2pt solid $cyan2

	.TodoBorderOff
		border: none
