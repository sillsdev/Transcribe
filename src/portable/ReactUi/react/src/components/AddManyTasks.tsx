import * as React from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import * as actions from '../actions/taskActions';
import { IUserSettingsStrings } from '../model/localize';
import { IProjectSettingsStrings } from '../model/localize';
import { IState } from '../model/state';
import uiDirection from '../selectors/direction';
import userStrings from '../selectors/localize';
import './AddManyTasks.sass';
import BackLink from './controls/BackLink';
import NextAction from './controls/NextAction';

interface IProps extends IStateProps, IDispatchProps {
};

export class AddManyTasks extends React.Component<IProps, object> {

    public AddManyTasks = () => {
        const { addManyTasks, selectedUser, selectedProject } = this.props;
        addManyTasks(selectedUser, selectedProject);
    }

    public doFetch = () => {
        const {fetchTasksOfProject, selectedProject} = this.props;
        fetchTasksOfProject(selectedProject);
    }

    public render() {
        const { direction, strings, strings2 } = this.props;
        const reactStringReplace = require('react-string-replace');

        const links = [ this.batchUploadSpreadsheetHelp, this.ourNamingConventionHelp ];
        const hotText = [ strings2.batchUpload, strings2.ourNamingConvention ];

        return (
            <div className={"AddManyTasks " + (direction && direction === "rtl" ? "rtl" : "ltr")}>
                <div className="panel">
                    <div className="titleRow">
                        <BackLink action={this.doFetch} target="/ProjectSettings" />
                    </div>
                    <div className="browseDiv">
                        <NextAction text={strings.browse.toUpperCase()} target={this.AddManyTasks} type="primary" />
                    </div>
                    <div className="helpText">
                        <span>
                            {reactStringReplace(strings2.learnToAddMany, /\{(\d+)\}/g, (match: string) => (
                                <a key={"a" + match} className="linkText" onClick={links[match]}>{hotText[match]}</a>
                            ))}
                        </span>
                    </div>
                </div>
            </div>
        );
    }

    private batchUploadSpreadsheetHelp = () => {
        this.props.showHelp("Concepts/Spreadsheet_convention.htm")
    }

    private ourNamingConventionHelp = () => {
        this.props.showHelp("Concepts/Folder_and_Files.htm")
    }

}
interface IStateProps {
    direction: string;
    selectedProject: string;
    selectedUser: string;
    strings: IUserSettingsStrings;
    strings2: IProjectSettingsStrings;
};

const mapStateToProps = (state: IState): IStateProps => ({
    direction: uiDirection(state),
    selectedProject: state.tasks.selectedProject,
    selectedUser: state.users.selectedUser,
    strings: userStrings(state, { layout: "userSettings" }),
    strings2: userStrings(state, {layout: "projectSettings"}),
});
interface IDispatchProps {
    addManyTasks: typeof actions.addManyTasks,
    fetchTasksOfProject: typeof actions.fetchTasksOfProject,
    showHelp: typeof actions.showHelp,
};

const mapDispatchToProps = (dispatch: any): IDispatchProps => ({
    ...bindActionCreators({
        addManyTasks: actions.addManyTasks,
        fetchTasksOfProject: actions.fetchTasksOfProject,
        showHelp: actions.showHelp,
    }, dispatch),
});

export default connect(mapStateToProps, mapDispatchToProps)(AddManyTasks);
