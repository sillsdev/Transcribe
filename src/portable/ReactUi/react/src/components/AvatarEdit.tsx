import * as React from 'react';
import AvatarEditor from 'react-avatar-editor';
import { connect } from 'react-redux';
import { Redirect } from 'react-router';
import { bindActionCreators } from 'redux';
import * as actions from '../actions/avatarActions';
import * as actions2 from '../actions/taskActions';
import { IProjectSettingsStrings, IUserSettingsStrings } from '../model/localize';
import { IState } from '../model/state';
import uiDirection from '../selectors/direction';
import userStrings from '../selectors/localize';
import './AvatarEdit.sass';
import BackLink from './controls/BackLink';
import IconButtonField from './ui-controls/IconButtonField';
import LabelCaptionUx from './ui-controls/LabelCaptionUx';

interface IAvatarEdit extends IAvatarState {
  discard: boolean;
  fromPath: string;
  imageData: any;
  scale: number;
}

interface IProps extends IStateProps, IDispatchProps {
  history: {
    location: {
      pathname: string;
    }
  },
  size?: number,
  snapShotTest?: boolean,
}

export class AvatarEdit extends React.Component<IProps, IAvatarEdit> {
  public state: IAvatarEdit;
  public editor: React.RefObject<AvatarEditor>;

  public constructor(props: IProps) {
    super(props);
    this.state = { ...this.props.avatar,
      discard: false,
      fromPath: "User",
      imageData: this.props.avatar.data,
      scale: 1
    };
    this.handleSave = this.handleSave.bind(this);
    this.handleDrop = this.handleDrop.bind(this);
    this.handleNewImage = this.handleNewImage.bind(this);
    this.handleScale = this.handleScale.bind(this);
    this.discard = this.discard.bind(this);
    this.editor = React.createRef();
  }

  public render() {
    const historyPath = this.props.history.location.pathname;
    const { size = 175, strings, strings2, direction } = this.props;
    const { borderRadius, discard, imageData, scale } = this.state;
    const backTo = historyPath.slice(0, historyPath.indexOf("/avatar"))
    const save = () => this.save(this);
    if (discard) {
      return (<Redirect to={backTo} />)
    }
    const wrapper = (this.props.snapShotTest) ? "Dropzone + Avatar": (
        <AvatarEditor
          ref={this.editor}
          scale={scale}
          width={size}
          height={size}
          rotate={0}
          borderRadius={size / (100 / borderRadius)}
          image={imageData}
          onImageChange={this.handleSave}
          onImageReady={this.handleSave}
          color={[128, 128, 128, 0.6]}
        />
    );

    const settingsStyle = direction? " " + direction: "";
    return (
      <div className={"AvatarEdit" + settingsStyle}>
        <div className="panel">
          <div className="titleRow">
            <BackLink action={save} target={backTo} />
            <div className="title">
              <LabelCaptionUx name={strings.editAvatar} type="H3" />
            </div>
          </div>
          <div className="data">
            <div>
              {wrapper}
            </div>
          </div>
          <div className="options">
            <div className="uploadContent">
              <LabelCaptionUx name={strings.newImage + " : "} />
              <button className="btn">{strings.browse + "..."}</button>
              <input
                name="newImage"
                type="file"
                onChange={this.handleNewImage}
              />
            </div>
            <div className="zoomContent">
              <LabelCaptionUx name={strings.zoom + " :"} />
              <input
                name="scale"
                type="range"
                onChange={this.handleScale}
                min="1"
                max="3"
                step="0.01"
                defaultValue="1"
              />
            </div>
          </div>
          <div className="action">
            <IconButtonField
              id="discard"
              caption={strings2.discardChanges}
              imageUrl="CancelIcon.svg"
              onClick={this.discard}
            />
          </div>
        </div>
      </div>
    );
  }

  private handleSave = () => {
    if (this.editor && this.editor.current){
      const img = this.editor.current.getImageScaledToCanvas().toDataURL();
      this.setState({ data: img })
    }
  };

  private handleDrop = (acceptedFiles: any) => {
    this.setState({ imageData: acceptedFiles[0], uri: acceptedFiles[0].name });
  };

  private handleNewImage = (e: any) => {
    this.setState({ imageData: e.target.files[0], uri: e.target.files[0].name });
  };

  private handleScale = (e: React.ChangeEvent<HTMLInputElement>) => {
    const scale = parseFloat(e.target.value);
    this.setState({ scale });
  };

  private discard() {
    this.setState({ ...this.state, discard: true })
  }

  private save(ctx: AvatarEdit) {
    const { avatar, saveAvatar, setSaveToProject, updateProjectAvatar, user } = this.props;
    const { data, saveToProject } = avatar;
    if (this.state.data !== data) {
      saveAvatar(this.state);
      if (saveToProject !== "") {
        updateProjectAvatar(user, saveToProject, {img: this.state.data});
        setSaveToProject("");
      }
    }
  }
}

interface IStateProps {
  avatar: IAvatarState;
  direction: string;
  strings: IUserSettingsStrings;
  strings2: IProjectSettingsStrings;
  user: string;
}

const mapStateToProps = (state: IState): IStateProps => ({
  avatar: state.avatar,
  direction: uiDirection(state),
  strings: userStrings(state, { layout: "userSettings" }),
  strings2: userStrings(state, { layout: "projectSettings" }),
  user: state.users.selectedUser
});

interface IDispatchProps {
  saveAvatar: typeof actions.saveAvatar;
  setSaveToProject: typeof actions.setSaveToProject;
  updateProjectAvatar: typeof actions2.updateProjectAvatar;
}

const mapDispatchToProps = (dispatch: any): IDispatchProps => ({
  ...bindActionCreators({
    saveAvatar: actions.saveAvatar,
    setSaveToProject: actions.setSaveToProject,
    updateProjectAvatar: actions2.updateProjectAvatar,
  }, dispatch)
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(AvatarEdit);
