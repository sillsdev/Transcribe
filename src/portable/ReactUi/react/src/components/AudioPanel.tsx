import * as React from 'react';
import * as actions from '../actions/audioActions';
import { log } from '../actions/logAction';
import * as actions1 from '../actions/taskActions';
import { ITranscriberStrings } from '../model/localize';
import './AudioPanel.sass';
import ControlPane from './ControlPane';
import EditorPane from './EditorPane';
import ProgressPane from './ProgressPane';

interface IProps extends IStateProps, IDispatchProps {
};

class AudioPanel extends React.Component<IProps, object> {
    public render() {
        log("AudioPanel")
        return (<div className="AudioPanel">
                    <div className="ProgressRow">
                        <ProgressPane {...this.props} />
                    </div>
                    <div className="EditorRow">
                        <EditorPane {...this.props} />
                    </div>
                    <div className="ControlRow">
                        <ControlPane {...this.props} />
                    </div>
                </div>
        )
    }
};

interface IStateProps {
    assignedReview: ITask[];
    assignedTranscribe: ITask[];
    direction: string;
    playing: boolean;
    playSpeedRate: number;
    selectedTask: string;
    selectedUser: string;
    jump: number;
    strings: ITranscriberStrings;
    users: IUser[];
};

interface IDispatchProps {
    playStatus: typeof actions.playStatus,
    playSpeedRateChange: typeof actions.playSpeedRateChange;
    jumpChange: typeof actions.jumpChange;
    reportPosition: typeof actions.reportPosition;
    saveTotalSeconds: typeof actions.saveTotalSeconds;
    setSubmitted: typeof actions.setSubmitted;
    showHelp: typeof actions1.showHelp,
};

export default AudioPanel;
