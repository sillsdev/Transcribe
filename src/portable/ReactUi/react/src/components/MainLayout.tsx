import * as React from 'react';
import { HotKeys } from 'react-hotkeys';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import * as actions from '../actions/audioActions';
import { log } from '../actions/logAction';
import * as actions2 from '../actions/taskActions';
import SucessPanel from '../components/SucessPanel';
import { ITranscriberStrings } from '../model/localize';
import { IState } from '../model/state';
import taskList from '../selectors';
import uiDirection from '../selectors/direction';
import userStrings from '../selectors/localize';
import currentProject from '../selectors/project';
import projectState from '../selectors/projectState';
import taskValues from '../selectors/taskValues';
import AudioPanel from './AudioPanel';
import GreetingPanel from './GreetingPanel';
import './MainLayout.sass';
import NavPanel from './NavPanel';
import TaskPanel from './TaskPanel';

interface IProps extends IStateProps, IDispatchProps {
    history: {
        location: {
            pathname: string;
        }
    }
};

const initialState = {
    highlightSeconds: 0,
}

class MainLayout extends React.Component<IProps, any> {
    private back: string;
    private faster: string;
    private forward: string;
    private playPause: string;
    private slower: string;
    private interval: any;

    public constructor(props: IProps) {
        super(props);
        this.state = {...initialState}
    }
    public componentWillMount()
    {
        const { selectedUser, users } = this.props;
        const user = users.filter(u => u.username.id === selectedUser)[0];

        // Get the hotkeys specified for the user
        if (user && user.hotkey !== undefined){
            this.playPause = user.hotkey.filter(h => h.id === "play-pause")[0].text.toLowerCase();
            this.back = user.hotkey.filter(h => h.id === "back")[0].text.toLowerCase();
            this.forward = user.hotkey.filter(h => h.id === "forward")[0].text.toLowerCase();
            this.slower = user.hotkey.filter(h => h.id === "slower")[0].text.toLowerCase();
            this.faster = user.hotkey.filter(h => h.id === "faster")[0].text.toLowerCase();         
        }
    }

    public componentWillUpdate() {
        const { todoHighlight, setToDoHighlight } = this.props
        if ( todoHighlight && this.state.highlightSeconds === 1 ) {
            setToDoHighlight(false);
            this.setState({ highlightSeconds: 0 });
        }
    }

    public componentDidMount() {
        this.interval = setInterval(() => this.tick(), 1000);
    }

    public componentWillUnmount() {
        clearInterval(this.interval);
    }

    public tick() {
        const { todoHighlight } = this.props
        if(todoHighlight)
        {
            this.setState((prevState: any) => ({
                ...this.state,
                highlightSeconds: prevState.highlightSeconds + 1
            }));
        }

    }

    public render() {
        const { jumpChange, playing, playSpeedRate, playSpeedRateChange, playStatus, saved, submit,
            assignedReview, assignedTranscribe, availableReview,  availableTranscribe, direction, project, todoHighlight} = this.props;

        log("MainLayout")
        const keyMap = {
            backKey: this.back,
            fasterKey: this.faster,
            forwardKey: this.forward,
            playPauseKey: this.playPause,
            slowerKey: this.slower,
        }

        const handlers = {
            backKey: (event: any) => {
                jumpChange(-2);
            },
            fasterKey: (event: any) => {
                let speedUp = parseFloat(playSpeedRate.toString()) + 0.1;
                if(parseFloat(playSpeedRate.toString()) >= 2.0)
                {
                    speedUp = 2.0
                }
                playSpeedRateChange(speedUp);
            },
            forwardKey: (event: any) => {
                jumpChange(2);
            },
            playPauseKey : (event: any) => {
                playStatus(!playing);
            },
            slowerKey: (event: any) => {
                let speedDown =  parseFloat(playSpeedRate.toString()) - 0.1;
                if (parseFloat(playSpeedRate.toString()) <= 0.5) {
                    speedDown = 0.5
                }
                playSpeedRateChange(speedDown);
            },
        };

        let editorScreen = <GreetingPanel   {...this.props} />

        const claim = project && project.claim? project.claim: false;
        if (assignedReview.length + assignedTranscribe.length > 0 ||
            (claim && availableReview.length + availableTranscribe.length  > 0))
        {
            const sync = project && project.sync? project.sync: false;
            editorScreen = (submit && saved) ? <SucessPanel  {...this.props} sync={sync} /> : <AudioPanel {...this.props} />;
        }

        let settingsStyle = this.props.history.location.pathname.length > 5? " Modal": ""
        settingsStyle = direction? settingsStyle + " " + direction: settingsStyle;

        return (
            <HotKeys keyMap={keyMap} handlers={handlers}>
                <div  className={"MainLayout" + settingsStyle} tabIndex={1}>
                    <div className="NavCol">
                        <NavPanel {...this.props} />
                    </div>
                    <div className={todoHighlight? "TaskCol TodoBorderOn": "TaskCol TodoBorderOff"}>
                        <TaskPanel {...this.props} />
                    </div>
                    <div className="MainCol">
                        {editorScreen}
                    </div>
                </div>
            </HotKeys>
        )
    }
};

interface IStateProps {
    assignedReview: ITask[];
    assignedTranscribe: ITask[];
    availableReview: ITask[];
    availableTranscribe: ITask[];
    direction: string;
    heading: string;
    jump: number;
    playSpeedRate: number;
    playing: boolean;
    project: IProject;
    projectState: string | undefined;
    saved: boolean;
    selectedTask: string;
    selectedUser: string;
    strings: ITranscriberStrings;
    submit: boolean;
    users: IUser[];
    todoHighlight: boolean;
};

const mapStateToProps = (state: IState): IStateProps => ({
    assignedReview: taskList(state).assignedReview,
    assignedTranscribe: taskList(state).assignedTranscribe,
    availableReview: taskList(state).availableReview,
    availableTranscribe: taskList(state).availableTranscribe,
    direction: uiDirection(state),
    heading: taskValues(state).heading,
    jump: state.audio.jump,
    playSpeedRate: state.audio.playSpeedRate,
    playing: state.audio.playing,
    project: currentProject(state),
    projectState: projectState(state),
    saved: state.audio.saved,
    selectedTask: state.tasks.selectedTask,
    selectedUser: state.users.selectedUser,
    strings: userStrings(state, { layout: "transcriber" }),
    submit: state.audio.submit,
    todoHighlight: state.tasks.todoHighlight,
    users: state.users.users,
});

interface IDispatchProps {
    completeReview: typeof actions2.completeReview;
    completeTranscription: typeof actions2.completeTranscription;
    playStatus: typeof actions.playStatus;
    playSpeedRateChange: typeof actions.playSpeedRateChange;
    jumpChange: typeof actions.jumpChange;
    reportPosition: typeof actions.reportPosition;
    saveTotalSeconds: typeof actions.saveTotalSeconds;
    setSubmitted: typeof actions.setSubmitted;
    assignTask: typeof actions2.assignTask;
    unassignTask: typeof actions2.unAssignTask;
    setToDoHighlight: typeof actions2.setToDoHightlight;
};

const mapDispatchToProps = (dispatch: any): IDispatchProps => ({
    ...bindActionCreators({
        assignTask: actions2.assignTask,
        completeReview: actions2.completeReview,
        completeTranscription: actions2.completeTranscription,
        jumpChange: actions.jumpChange,
        playSpeedRateChange: actions.playSpeedRateChange,
        playStatus: actions.playStatus,
        reportPosition: actions.reportPosition,
        saveTotalSeconds: actions.saveTotalSeconds,
        setSubmitted: actions.setSubmitted,
        setToDoHighlight: actions2.setToDoHightlight,
        unassignTask: actions2.unAssignTask,
    }, dispatch),
});

export default connect(mapStateToProps, mapDispatchToProps)(MainLayout);
