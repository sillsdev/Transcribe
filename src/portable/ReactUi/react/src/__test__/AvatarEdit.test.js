/* jshint esversion: 6 */
import React from 'react';
import expect from 'expect';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { MemoryRouter as Router} from 'react-router-dom';
import renderer from 'react-test-renderer';
import sinon from 'sinon';
import {AvatarEdit} from '../components/AvatarEdit';

configure({ adapter: new Adapter() });

// Snapshot for AvatarEdit
describe('>>>Control: AvatarEdit --- Snapshot',()=>{
    const minProps = {
        avatar: {
            data: "/api/images/smile.svg",
            uri: "/api/images/smile.svg"
        },
        direction: "ltr",
        history: {
            location: {
                pathname: "/main/avatar"
            }
        },
        strings: {
            browse: "Browse",
            newImage: "New Image File",
            zoom: "Scale",
        },
        strings2: {
            discardChanges: "Discard Changes"
        }
    };

    it('+++capturing Snapshot of AvatarEdit', () => {
        const renderedValue =  renderer.create(<Router><AvatarEdit {...minProps} snapShotTest={true}/></Router>).toJSON()
        expect(renderedValue).toMatchSnapshot();
    });
});
//*******************************************************************************************************
describe('>>>Control: AvatarEdit', () => {
    let wrapper;
    const minProps = {
        avatar: {
            data: "/api/images/smile.svg",
            uri: "/api/images/smile.svg"
        },
        direction: "ltr",
        history: {
            location: {
                pathname: "/main/avatar"
            }
        },
        strings: {
            browse: "Browse",
            newImage: "New Image File",
            zoom: "Scale",
        },
        strings2: {
            discardChanges: "Discard Changes"
        }
    };

    beforeEach(() => {
        wrapper = shallow(<AvatarEdit {...minProps} />);
    })

    it('+++ renders without exploding', () => {
        expect(wrapper.length).toEqual(1);
    });
});
