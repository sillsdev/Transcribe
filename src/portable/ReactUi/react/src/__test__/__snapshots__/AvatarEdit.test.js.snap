// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`>>>Control: AvatarEdit --- Snapshot +++capturing Snapshot of AvatarEdit 1`] = `
<div
  className="AvatarEdit ltr"
>
  <div
    className="panel"
  >
    <div
      className="titleRow"
    >
      <div
        className="BackLink"
        id="BackLink"
      >
        <a
          href="/main"
          onClick={[Function]}
        >
          ‚ùÆ
        </a>
      </div>
      <div
        className="title"
      >
        <span
          className="LabelCaptionUx H3 label label-default hidden"
        />
      </div>
    </div>
    <div
      className="data"
    >
      <div>
        Dropzone + Avatar
      </div>
    </div>
    <div
      className="options"
    >
      <div
        className="uploadContent"
      >
        <span
          className="LabelCaptionUx  label label-default"
        >
          New Image File : 
        </span>
        <button
          className="btn"
        >
          Browse...
        </button>
        <input
          name="newImage"
          onChange={[Function]}
          type="file"
        />
      </div>
      <div
        className="zoomContent"
      >
        <span
          className="LabelCaptionUx  label label-default"
        >
          Scale :
        </span>
        <input
          defaultValue="1"
          max="3"
          min="1"
          name="scale"
          onChange={[Function]}
          step="0.01"
          type="range"
        />
      </div>
    </div>
    <div
      className="action"
    >
      <div
        className="IconButtonField"
        id="discard"
        onClick={[Function]}
      >
        <img
          alt="D"
          className="CaptionImage"
          src="/assets/CancelIcon.svg"
        />
        <label
          className="Caption"
        >
          DISCARD CHANGES
        </label>
      </div>
    </div>
  </div>
</div>
`;
